engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -339,8 +340,12 @@ Object doWarning(
 
     arguments[thisArgumentPosition] = selfWithoutWarnings;
 
-    Object result = childDispatch.execute(frame, state, symbol, selfWithoutWarnings, arguments);
-    return WithWarnings.appendTo(EnsoContext.get(this), result, arrOfWarnings);
+    try {
+      Object result = childDispatch.execute(frame, state, symbol, selfWithoutWarnings, arguments);
+      return WithWarnings.appendTo(EnsoContext.get(this), result, arrOfWarnings);
```

Comment:
There is following code above this change:
```java
    if (childDispatch == null) {
      CompilerDirectives.transferToInterpreterAndInvalidate();
      Lock lock = getLock();
      lock.lock();
      try {
        if (childDispatch == null) {
          childDispatch =
              insert(
                  build(
                      invokeFunctionNode.getSchema(),
                      invokeFunctionNode.getDefaultsExecutionMode(),
                      invokeFunctionNode.getArgumentsExecutionMode(),
                      thisArgumentPosition));
          childDispatch.setTailStatus(getTailStatus());
          childDispatch.setId(invokeFunctionNode.getId());
          notifyInserted(childDispatch);
        }
      } finally {
        lock.unlock();
      }
    }
```

1. it is wrong: we should create a factory method and use `@Cached` - I don't think there is a reason for that not to work
2. it sets `setTailStatus`

Shouldn't we have two `@Specialization`s - one for values `WithWarnings` and one for without warnings and just make sure the `TailStatus` for `WithWarnings` is `NOT_TAIL`?

Reply:
I can explore that solution with another specialization.

Reply:
> one for values WithWarnings and one for without warnings

This specialization has `hasWarnings(self)` guard so the second option is not possible.

Reply:
I don't get it. What is not possible?

Reply:
My patch looks exactly the same. What I meant by it is that I initially understood your " one for values WithWarnings and one for without warnings " comment as we should split the existing `doWarning` into two cases. But based on your comment, this is not what you meant. 

> There could be a problem with instrumentation, but InvokeMethodNode isn't instrumentable. 

I believe that's the primary reason why we don't use `@Cached` in this and other places. See the 
```
// Cannot use @Cached for childDispatch, because we need to call notifyInserted.
```
comment.

Reply:
`notifyInserted`! Not `insert`. I see.

Reply:
> we should create a factory method and use @Cached - I don't think there is a reason for that not to work

OK, so we know that it doesn't work because of `WarningInstrumentatiopnTest`. Sorry for false hope request.

Reply:
> based on your comment, this is not what you meant.

Didn't I? What is wrong with:

> Shouldn't we have two @Specializations - one for values WithWarnings and one for without warnings and just make sure the TailStatus for WithWarnings is NOT_TAIL?

It still solves the problem of checking warnings all the time.

Reply:
After [5628c71](https://github.com/enso-org/enso/pull/6849/commits/5628c711997ae3b0de0eb6f36e5ac4011660b15e) I see just three changes setting `NOT_TAIL`:
- [InvokeCallableNode](https://github.com/enso-org/enso/pull/6849/files#diff-01d58f3323980e95f4bf73d7c122e1633dafc4f85fa49661106b82755f0aeaa4R280)
- [InvokeConversionNode](https://github.com/enso-org/enso/pull/6849/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R331)
- [InvokeMethodNode](https://github.com/enso-org/enso/pull/6849/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R331)

That seems like the right fix.